src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -0,0 +1,185 @@
+#pragma warning disable IDE0051
+
+using Neo.Cryptography;
+using Neo.Cryptography.ECC;
+using Neo.Ledger;
+using Neo.Network.P2P.Payloads;
+using Neo.Persistence;
+using Neo.SmartContract.Manifest;
+using Neo.VM.Types;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+
+namespace Neo.SmartContract.Native.Oracle
+{
+    public sealed partial class OracleContract : NativeContract
+    {
+        private const int MaxUrlLength = 256;
+        private const int MaxFilterLength = 128;
+        private const int MaxCallbackLength = 32;
+        private const int MaxUserDataLength = 512;
+
+        private const byte Prefix_NodeList = 8;
+        private const byte Prefix_RequestId = 9;
+        private const byte Prefix_Request = 7;
+        private const byte Prefix_IdList = 6;
+
+        private const long OracleRequestPrice = 0_50000000;
+
+        public override int Id => -4;
+        public override string Name => "Oracle";
+
+        internal OracleContract()
+        {
+            Manifest.Features = ContractFeatures.HasStorage;
+        }
+
+        [ContractMethod(0, CallFlags.AllowModifyStates)]
+        private void Finish(ApplicationEngine engine)
+        {
+            Transaction tx = (Transaction)engine.ScriptContainer;
+            OracleResponse response = tx.GetAttribute<OracleResponse>();
+            OracleRequest request = GetRequest(engine.Snapshot, response.Id);
+            StackItem userData = BinarySerializer.Deserialize(request.UserData, engine.MaxStackSize, engine.MaxItemSize, engine.ReferenceCounter);
+            engine.CallFromNativeContract(null, request.CallbackContract, request.CallbackMethod, request.Url, userData, (int)response.Code, response.Result);
```

Comment:
What happens if the callback withdraw gas from the `Oracle Contract`? We should force `CalledByEntry` in `OracleResponse.Verify` https://github.com/neo-project/neo/pull/1738/files#diff-1f630fc327408ee85b57157bc54326f3R68

Reply:
This situation can't happen. Only reponse tx can take oracle contract as sender.

Reply:
`ResponseTx` call your request callback, if the scope of the signature is not well configured, it could happen.

Reply:
Why `CalledByEntry`? Currently it is `WitnessScope.None`, which means it only pay for the fees.

Reply:
Yes, https://github.com/neo-project/neo/pull/1738/files#diff-1f630fc327408ee85b57157bc54326f3R68 it's good, my mistake :)